src/dmd/utils.d

```diff
@@ -138,3 +139,44 @@ const(char)[] toDString (const(char)* s) pure nothrow @nogc
 {
     return s ? s[0 .. strlen(s)] : null;
 }
+
+/**
+   Copy the content of `src` into a C-string ('\0' terminated) then call `dg`
+
+   The intent of this function is to provide an allocation-less
+   way to call a C function using a D slice.
+   The function internally allocates a buffer if needed, but frees it on exit.
+
+   Note:
+     The argument to `dg` is `scope`. To keep the data around after `dg` exits,
+     one has to copy it.
+
+   Params:
+     src = Slice to use to call the C function
+     dg  = Delegate to call afterwards
+
+   Returns:
+     The return value of `T`
+*/
+auto toCStringThen(alias dg)(const(char)[] src) nothrow
```

Comment:
Maybe include "temporary" in the name, s.t. it's always clear that keeping the C string around is dangerous?

Reply:
I used this name because it matches [what's already there](
https://github.com/dlang/dmd/blob/c3f6320d59ec14d6fa81a18f92b59393671b346a/src/dmd/root/filename.d#L818), namely `toWStringzThen` and `toExtendedPathThen`.
Marking those functions `@safe` should enforce that `scope` is checked for, but we cannot (yet) do that.

Reply:
>  should enforce that scope is checked for, but we cannot (yet) do that.

Well safety is still broken for now too, e.g. https://issues.dlang.org/show_bug.cgi?id=19175